### PR TITLE
K01: Implement Full Validate Profile Method & Required Methods

### PR DESCRIPTION
**Pull request by christopher-davis-afs**
## Describe your changes

This PR finishes fleshing out the profile validation work in `SmartChargingHandler`. It finalizes the implementation for a few functional requirements:

* K01.FR.06
* K01.FR.38
* K01.FR.03

With this PR, we can now call `validate_profile()` and, based on the type of the given profile, run the specific validation methods that are applicable. We also simplify the function signatures of some of the previously added methods to take the EVSE ID.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

**Comment by christopher-davis-afs** [comment]
CC: @shankari

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [inline lib/ocpp/v201/smart_charging.cpp:374]: Please replace these by either:
```
profile.validFrom = profile.validFrom.value_or(ocpp::DateTime()));
```
or
```
if (!profile.validFrom.has_value()) {
    profile.validFrom = ocpp::DateTime();
}
```

To me that would be more clear what the intention is. Also `DateTime()` already default constructs to `date::utc_clock::now()` ;)

**Review by marcemmers — COMMENTED** [review]
Some comments before this is an actual PR. Did not look at the tests in its entirety yet.

File: include/ocpp/v201/smart_charging.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/smart_charging.hpp:57]: Should we make those changes to `charge_point` here?

File: include/ocpp/v201/smart_charging.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/smart_charging.hpp:73]: If we do that, we feel like we would lose the ability to easily unit test and lose the high level of coverage that we currently have. It would also make it difficult to test-drive functionality that remains to be implemented. Would it be acceptable to make them `protected` instead of `private`?

File: include/ocpp/v201/smart_charging.hpp
Comment (marcemmers) [reply include/ocpp/v201/smart_charging.hpp:57]: Its up to you. If you make a small PR for this I can approve it quickly but you can also put it in this one.

File: include/ocpp/v201/smart_charging.hpp
Comment (couryrr-afs) [reply include/ocpp/v201/smart_charging.hpp:73]: This was brought up on the General Q&A call and it does not appear that the community has any concern with `protected` methods.

Comment (marcemmers) [reply include/ocpp/v201/smart_charging.hpp:73]: If that is the best option then I have no problem with that. Just felt strange having them exposed while the user of the class should not use them.

Comment (couryrr-afs) [reply include/ocpp/v201/smart_charging.hpp:73]: @marcemmers it is a reasonable request. We just want to balance the ability to test and locking the interface down.

File: include/ocpp/v201/smart_charging.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/smart_charging.hpp]: Decided to do it here.

File: include/ocpp/v201/smart_charging.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/smart_charging.hpp:75]: Marked them all as protected

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (marcemmers) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:406]: I would suggest to make the UUID a definition as well instead of generating one each time. Just to limit changes when running the tests.

**Review by marcemmers — CHANGES_REQUESTED** [review]
Next batch of comments. I now looked through all the code so I don't expect any more than these.

I like the clear unit tests being really specific :+1:

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (marcemmers) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:185]: I've created a PR here: https://github.com/EVerest/libocpp/pull/664

Let me know if this would be usable for you or not. I had to change to a `shared_ptr` for the `EvseInterface` but I have an idea on how to remove the need for passing the map by reference as well, eliminating the `shared_ptr` too.

Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:185]: Ultimately I think that would be great, but the issue we ran into was that `get_value()`, `get_optional_value()`, and `request_value()` return templated types, and I could not find a way to mock those.

How would I request a variable using that PR?

Comment (marcemmers) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:185]: I found that issue as well and had to work around it a little bit. If you want to mock the template function `get_value()` for example, you could use `expect_request_value()` to set the return value.

This is possible since all the templated functions only use a call to `request_value_internal()` that is mockable. `expect_request_value()` is just a helper to simplify the usage a bit.

there are probably a lot of things that can still be improved or added but it should be a good starting point I think.

Comment (marcemmers) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:185]: For example this should do the trick:
```
this->device_model.expect_request_value(ControllerComponentVariables::ChargingStationSupplyPhases, "5");

EXPECT_EQ(this->device_model.get_value<int32_t>(ControllerComponentVariables::ChargingStationSupplyPhases), 5);
```

Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:166]: Talking this through with the team, I think we'd prefer to keep it as is. Using the real database and loading it in-memory gives us free testing of real state, and gives us code that's easier to parse than using a mock.

**Comment by shankari** [comment]
@the-bay-kay the high-level description is not quite correct, this PR addresses the following functional requirements:
- K01.FR.06
- K01.FR.20
- K01.FR.26
- K01.FR.38
- K04.FR.03

Please read through my review and review the changes yourself so you get familiar with the process.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:386]: why do you need this? the `DEFAULT_PROFILE_ID. = 1` which is the same as the `DEFAULT_EVSE_ID`. So just validating against `DEFAULT_PROFILE_ID` should give the same result and potentially be less confusing because we validate against the same ID as the profile

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:377]: similarly, not sure why the `+1` was needed here given that `DEFAULT_PROFILE_ID` is already 1

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:374]: we need an FR here or an explanation for why it is not relevant to any one FR

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:604]: why does this now need a datetime when it didn't before?

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:745]: spelling

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:754]: I'm curious as to why we use `utc_clock::time_point::max/min` when we want to say "undefined" in `install_profile_on_evse` but `{}` in `create_charging_profile`. It seems like it would be much easier to keep it consistent.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:837]: what is invalid about this? please add a comment explaining why...

**Review by shankari — APPROVED** [review]
For ease of review, it would be helpful to have a description of which functions are related to with FR, preferably in the code, so that it is easier for future maintainers to find it while reading through the code.

I have several comments, but they are primarily around improved documentation and design choices. I don't see anything obviously incorrect, and obviously the tests pass. This has already been pending for a while, so I will approve and anticipate that the comments will be addressed in a future cleanup PR.

File: lib/ocpp/v201/charge_point.cpp
Comment (marcemmers) [reply lib/ocpp/v201/charge_point.cpp:83]: This was by my suggestion since that would be preferred over passing a reference to a `unique_ptr` into the `SmartChargingHandler`. Although I have been in some discussions, and would have to agree with them, that for most of these we don't need to pass the ptr at all and could just pas a reference to the `DeviceModel` in this case.

File: include/ocpp/v201/smart_charging.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/smart_charging.hpp:75]: They have to:

> If a profile does not have validFrom or validTo set, we conform the values to a representation that fits the spec.

And in `validate_profile_schedules()`:

> If a profiles charging schedule period does not have a valid numberPhases, we set it to the default value (3).

Comment (marcemmers) [reply include/ocpp/v201/smart_charging.hpp:75]: I have also noted this issue before and still agree with @shankari here. I think it should be possible to separate the responsibilities a bit. Maybe we can already update those fields before calling the validate functions? Would make it a lot clearer what is going on, something along the lines of: `set_default_number_of_phases()` and `set_default_valid_from_to_times()`

File: lib/ocpp/v201/smart_charging.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/smart_charging.cpp:61]: These aren't intended to be sent in the response - rather, they're meant to provide detailed information for logging and for providing a detailed response at the message handling level.

File: lib/ocpp/v201/smart_charging.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/smart_charging.cpp:89]: Yes, we get multiple values from the device model already:

```c++
auto supported_charging_rate_units = this->device_model->get_value<std::string>(ControllerComponentVariables::ChargingScheduleChargingRateUnit);
```

```c++
!device_model->get_optional_value<bool>(ControllerComponentVariables::ACPhaseSwitchingSupported).value_or(false)
```

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:167]: This method sets the defaults so that all of the other test cases pass. For the ACPhaseSwitchingSupported tests, we need to do this as we can't make the value blank after it's already been initialized with a valid value. So we can't do what we do with e.g. the FR26 tests.

As written, this function allows us to initialize a device model with defaults that will work for all cases, including the case where we want to test our code for when the value is missing.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:756]: `install_profile_on_evse()` bypasses our validation/conforming functions, so we need to store the profile as if it was already modified for storage and later checks. In contrast, `create_charging_profile()` represents actual data that may be passed into `validate_profile()`. Thus, the values are missing instead of being set to `min()` or `max()`.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:328]: Added a test that should cover this case.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:401]: This is just an explicit name for this test (testing whether or not the EVSE ID is greater than zero) in lieu of a comment.

File: include/ocpp/v201/smart_charging.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/smart_charging.hpp:80]: We have a note on this. We will change this in our add profile branch in order to reduce the amount of work done rebasing.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp]: We added a variable `extraneous_start_schedule`, to clarify the issue within this test.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:389]: This is a simple pre-check for checking that we are passing in the right type of profile. This is not associated with any functional requirement. We think this code is self-documenting.

File: doc/ocpp_201_status.md
Comment (christopher-davis-afs) [reply doc/ocpp_201_status.md:1160]: I think this comment highlights the outcome of a conversation, but I don't think incoming contributors would need to know the context of the discussion in order to understand this outcome.

File: include/ocpp/v201/smart_charging.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/smart_charging.hpp:75]: We understand the idea of using `const`, but this allows us to set the defaults required by the specification while keeping the public footprint minimal. As-is, `validate_profile()` and `add_profile()` can be called directly without needing to know about the internals of the profiles.

Comment (marcemmers) [reply include/ocpp/v201/smart_charging.hpp:75]: Of course I understand that you do ;) I just think this is not ideal. On the other hand. it would also be a small update down the line if we ever decide to change this so I won't let this block the PR.

**Review by marcemmers — APPROVED** [review]

**Comment by marcemmers** [comment]
@christopher-davis-afs Just give me a sign that it is ready to merge and I will merge it ;)

**Comment by christopher-davis-afs** [comment]
It is ready :)

File: doc/ocpp_201_status.md
Comment (shankari) [reply doc/ocpp_201_status.md:1160]: I disagree. "makes no sense with the rest of the spec" is not an detailed outcome. **Why** does it make no sense? I am not asking for much, just a link to the conversation so that conscientious contributors **can** read the context if they choose to do so. If you still want to push back on this suggestion, I would like some backup for your thoughts 😄

File: include/ocpp/v201/smart_charging.hpp
Comment (shankari) [reply include/ocpp/v201/smart_charging.hpp:98]: Just noting that this needs to be fixed in the cleanup PR as well

File: include/ocpp/v201/smart_charging.hpp
Comment (shankari) [reply include/ocpp/v201/smart_charging.hpp:75]: I agree with @marcemmers that while this is not important enough to hold up the PR (and I approved it as well), I would like to have this tracked as a cleanup change in a subsequent PR. When I read `validate_XXX`, I don't expect that it will modify the input parameters.

File: lib/ocpp/v201/charge_point.cpp
Comment (shankari) [reply lib/ocpp/v201/charge_point.cpp:83]: Right, that is what I thought - why not just pass a reference? Again, I would suggest filing a cleanup issue, and bringing it up for discussion for future cleanup. Even if we don't make any changes, at least we should do so intentionally 😄

File: lib/ocpp/v201/smart_charging.cpp
Comment (shankari) [reply lib/ocpp/v201/smart_charging.cpp:61]: ok; can you clarify when/where these will actually be sent in the response? Ideally, we would ensure that the errors are standardized, since OCPP does not define status codes or strings.

File: lib/ocpp/v201/smart_charging.cpp
Comment (shankari) [reply lib/ocpp/v201/smart_charging.cpp:192]: Ah I remembered it being changed from Evse to EvsetInterface in https://github.com/EVerest/libocpp/commit/ce792235f15cd2b75049a513ef234a8dcec13327 But it wasn't a change from `evse_id`

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:170]: To be more explicit, my feedback is that `create_device_model` should take in a single argument, which is the path. It should use defaults for everything else. Each test should then override the parts of the device model that are relevant to itself.

Concretely, I don't see why `ac_phase_switching_supported` is special enough that it needs to be an argument but (say) `charging_rate_unit_cv.component` is not.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:401]: My comment is not about the test name but the variable `const int EVSE_ID_1`

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:389]: This may be obvious to somebody who has spent the past several months steeped in this code, but it is certainly not self-documenting to somebody who is encountering this code for the first time. Please add a one-line comment that says exactly what you said above "This is a simple pre-check for checking that we are passing in the right type of profile"

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:768]: Please add this as a comment to the source code as well

**Comment by shankari** [comment]
@couryrr-afs @drmrd @wjmp I believe that there were several comments that here that the reviewers wanted to change, but didn't want to hold up the merge for. Now that the merge is done, it would be great to see a cleanup PR address the comments. Also, some of my comments were marked as resolved, but I don't agree with the resolution 😄

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp]: I'm sorry, but it is still not clear to me. Before the `create_charging_schedule` function is called, there are no profiles. This adds a profile and sets it. The arguments to `create_charging_profile` appear to be identical to all the other calls, including ones that are valid. What is different here? Is there a specific FR that you can point to for the reader to get some additional context....

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:401]: Yes, that's what I meant. This is just an explicit name for the variable that's relevant to this test.

**Comment by couryrr-afs** [comment]
> @couryrr-afs @drmrd @wjmp I believe that there were several comments that here that the reviewers wanted to change, but didn't want to hold up the merge for. Now that the merge is done, it would be great to see a cleanup PR address the comments. Also, some of my comments were marked as resolved, but I don't agree with the resolution 😄

@shankari  Yep — we haven’t forgotten about the code items and can get the comments added to the code base you requested. We have been prioritizing getting the Validate Profile PR merged, the Add Profiles PR ready in support of Alfen’s timelines and some other blocker around that. I thought it best to resolve comments prior to presenting the PR to the community as to avoid any confusion or further delay based on your comment about the review.
 
We will have a PR out soon.

File: lib/ocpp/v201/smart_charging.cpp
Comment (couryrr-afs) [reply lib/ocpp/v201/smart_charging.cpp:61]: The use of these are only for internal tracing. The code that will leverage them is in progress. It will be in the handler for `SetChargingProfileRequest` the goal was to simply allow for logging so that when an issue arises the community could easily find where a profile was failing ie if the result of validation failed with `ChargingSchedulePeriodPhaseToUseACPhaseSwitchingUnsupported` you simply have to search where that is set.

As for a standardized set of these code we were adding them as we did our development with an internal use mentality. We can bring this up in a working group if that would help but it does not seem like something that should be formally created.

Comment (shankari) [reply lib/ocpp/v201/smart_charging.cpp:61]: > We can bring this up in a working group if that would help
Yes, I think it would be worthwhile to bring this up in a working group, if only to figure out what the plans are

> but it does not seem like something that should be formally created.
The MREC project, also funded by JOET, is completely focused on formally creating such standards. We should make sure to incorporate their recommendations.

**Pull request merged**
